vncviewer/EmulateMB.cxx

```diff
@@ -194,23 +205,198 @@ static const signed char stateTab[11][5][3] = {
+    {  0,  0,  6 },   /* other button -> self */
+    {  0,  0,  6 }    /* key pressed -> self */
+  },
+/* 7 after emulation, button up, mod down. mod not sent to server */
+  {
+    { 10,  0,  0 },   /* mod down -> error, ground */
+    {  0,  0,  0 },   /* mod up -> ground */
+    {  0,  1,  3 },   /* left button -> emulating both down */
+    {  1,  0,  6 },   /* other button -> non emulated, mod physicall down. button up */
+    {  1,  0,  6 }    /* key pressed -> non emulated, mod physicall down. button up */
+  }
+};
+
+EmulateMB::EmulateMB(rdr::U32 emulateMBModKey)
+  : state(0), emulatedButtonMask(0), timer(this), emulateMiddleButtonModifierKey(emulateMBModKey),
+    mbState(0), mbLastButtonMask(0)
+{
+}
+
+void EmulateMB::mbFsmDoMod(int modAction)
 {
+  if (modAction == 1 || modAction == 2)
+    writeKeyEvent(emulateMiddleButtonModifierKey, modKeyCode, true);
+  if (modAction == -1 || modAction == 2)
+    writeKeyEvent(emulateMiddleButtonModifierKey, modKeyCode, false);
+}
+
+bool EmulateMB::filterKeyPressRelease(bool is_press, int keyCode, rdr::U32 keySym)
+{
+  if (!emulateMiddleButtonMod)
+    return false;
+
+  Action *action;
+  if (keySym == emulateMiddleButtonModifierKey || keyCode == modKeyCode) {
+    if (is_press) {
+      action = &mbStateTab[mbState][0];
+      modKeyCode = keyCode;
+    }
+    else
+      action = &mbStateTab[mbState][1];
+  }
+  else {
+    action = &mbStateTab[mbState][4];
+  }
+
+  mbFSM(*action);
+  // If another key was pressed then signal the caller to send it also.
+  return keyCode == modKeyCode;
+}
+
+bool EmulateMB::filterKeyPress(int keyCode, rdr::U32 keySym) {
+  return filterKeyPressRelease(true, keyCode, keySym);
+}
+
+bool EmulateMB::filterKeyRelease(int keyCode)
+{
+  return filterKeyPressRelease(false, keyCode, -1);
+}
+
+void EmulateMB::mbFSM(Action action)
+{
+  if (action[0] == 10) {
+    vlog.error(_("Bad state transition. In state %d\n"), mbState);
+    mbState = 0;
+    return;
+  }
+  mbFsmDoMod(action[0]);
+  mbState = action[2];
+}
+
+void EmulateMB::filterPointerEventMod(const rfb::Point& pos, int buttonMask)
+{
+  Action *action;
+
+  if (buttonMask != mbLastButtonMask) {
+    // Button 1 changed.
+    if ((buttonMask & 1) != (mbLastButtonMask & 1))
+      action = &mbStateTab[mbState][2]; // Left button
+    else
+      action = &mbStateTab[mbState][3]; // Other button
+
+    mbLastButtonMask = buttonMask;
+    if ((*action)[1]) {
+      // Emulate. Set button 2 to equal button 1 and turn off button 1.
+      int button1 = buttonMask & 0x1;
+      buttonMask = (buttonMask & ~0x3) | button1 << 1;
+    }
+
+    mbFSM(*action);
+  }
+  sendPointerEvent(pos, buttonMask);
 }
 
 void EmulateMB::filterPointerEvent(const rfb::Point& pos, int buttonMask)
+{
+  if (emulateMiddleButton)
+    filterPointerEventLR(pos, buttonMask);
+  else if (emulateMiddleButtonMod)
+    filterPointerEventMod(pos, buttonMask);
+  else
+    sendPointerEvent(pos, buttonMask);
```

Comment:
This is really nice! Keeps things neat and not having to solve both modes in the same code.